src/classes/BDI_AdditionalObjectService.cls

```diff
@@ -0,0 +1,391 @@
+                        } else { 
+                            dataImportService.LogBDIError(objWrap.di, result.getErrors()[0].getMessage(),UTIL_Namespace.adjustNPSPNamespace(objWrap.objMapping.Imported_Record_Status_Field_Name__c));
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /*******************************************************************************************************
+    * @description Determines if there is a custom mapping logic class and constructs it if there is, otherwise returns the default
+    * mapping class.
+    * @param objMapping The Data Import Object Mapping to look for a custom mapping logic class on.
+    */
+    private BDI_ObjectMappingLogic getObjectMappingLogicClass (Data_Import_Object_Mapping__mdt objMapping) {
+        BDI_ObjectMappingLogic objMappingLogicClass;
+
+        if (objMapping.Custom_Mapping_Logic_Class__c != null) {
+            Type custLogicClassType = Type.forName(UTIL_Namespace.adjustNPSPNamespace(objMapping.Custom_Mapping_Logic_Class__c));
+            objMappingLogicClass = (BDI_ObjectMappingLogic)custLogicClassType.newInstance();
+        } else {
+            objMappingLogicClass = new BDI_ObjectMappingLogic();
+        }
+        return objMappingLogicClass;
+    }
+
+
+    /*******************************************************************************************************
+    * @description Determines if the predecessor to this Data Import Object Mapping exists
+    * @param predecessor The Data Import Object Mapping for which the imported field can be checked to see if it exists.
+    */
+    private Boolean predecessorExists(BDI_ObjectWrapper objWrap, Data_Import_Object_Mapping__mdt predecessor){
+        return objWrap.di.get(UTIL_Namespace.adjustNPSPNamespace(predecessor.Imported_Record_Field_Name__c)) != null;
+    }
+
+
+    /*******************************************************************************************************
+    * @description Validates the field mappings to make sure that the required ones are not null and that at least one non imported record
+    * field is populated.
+    * @param objWrap The BDI_ObjectWrapper that should have its field mappings validated.
+    */
+    private Boolean validateFields(BDI_ObjectWrapper objWrap){
+
+        Boolean result = false;
+        Boolean anyNonImportedRecordFieldsPopulated = false;
+        Boolean allRequiredFieldsPopulated = true;
+
+        String requiredFieldErrorMsg = System.label.bdiAdditionalObjRequiredFieldsNull;
+
+        // First determine if there is at least one field populated that was not automatically populated by the import of another object.
+        // This is done to ensure that the user actually intended to attempt the creation of this object.
+        for (Data_Import_Field_Mapping__mdt fieldMapping : objWrap.fieldMappings) {
```

Comment:
I got tripped up by the duplication of the fieldMapping variable name in here and the class variable "BDI_FieldMappingCustomMetadata fieldMapping". Recognizing you may have renamed this one based on previous feedback, it would still be nice not to have that naming overlap

Reply:
updated to fieldMappingCustomMetadata